Reducing memory usage: removing useless logits computation in generate()

Cyrilvallez opened the pull request:
# What does this PR do?

This is the PR related to the discussion in #30860. 
I followed was has been done in `Jamba` and added support for the `num_logits_to_keep` argument in `forward()`. However, even if this argument is `None`, the logits will only be upcasted to float if `labels` are passed (in order to accurately compute the loss). Otherwise, the upcasting only happen in the `generate()` functions.

For now, I only modified `Llama` and `Mistral`, but if you agree on the changes I will add support for more models.

## Benchmarks

Here I provide some benchmarks of the peak memory usage. For each input size, I generated 10 additional tokens.
Of course, since for few additional tokens the memory peak scales only with the first forward pass (at least when computing the whole logits matrix), and that the first forward scales linearly with input size and batch size (with new attention algorithms), the gain is actually constant for all input sizes and generation methods (except for contrastive search, which artificially increase the batch size after the first forward, thus the memory usage is slightly different). However, I still provide results for all generation methods here for completeness.

Basically we get: 
Llama3 8B -> MIND-BLOWING 3.62 memory usage reduction factor (due to large vocabulary)
Llama2 7B -> 1.17 reduction factor
Mistral 7B -> 1.32 reduction factor

Note that the memory reduction shown here is **on top** of whatever gains #30536 already provides for small new additional tokens, as I am comparing memory with the main transformers branch after it was merged. It integrates very nicely with that last PR, as the last one was providing most benefits when generating  more tokens, and this one provides gains for small new number of tokens.

[greedy.pdf](https://github.com/user-attachments/files/15645353/greedy.pdf)
[sample.pdf](https://github.com/user-attachments/files/15645374/sample.pdf)
[beam sample.pdf](https://github.com/user-attachments/files/15645320/beam.sample.pdf)
[beam search.pdf](https://github.com/user-attachments/files/15645334/beam.search.pdf)
[group beam search.pdf](https://github.com/user-attachments/files/15645364/group.beam.search.pdf)
[contrastive search.pdf](https://github.com/user-attachments/files/15645344/contrastive.search.pdf)

Here is a link to the benchmark script: https://gist.github.com/Cyrilvallez/92f48e402aa2968c854a8128796f50c3

## Who can review?

@ArthurZucker @gante Let me know what you think about the proposed changes!

Cyrilvallez committed:
Add .float() in all generation methods logit outputs

Cyrilvallez committed:
Switch float-casting of logits to training only for main models

Cyrilvallez committed:
Add `num_logits_to_keep` in Llama and add it by default in generate

Cyrilvallez committed:
Apply style

Cyrilvallez committed:
Add num_logits_to_keep as arg in prepare_input_for_generation

Cyrilvallez committed:
Add support for Mistral

Cyrilvallez committed:
Revert models except llama and mistral

Cyrilvallez committed:
Fix default None value in _supports_num_logits_to_keep()

amyeroberts commented:
cc @ArthurZucker @gante

Cyrilvallez committed:
Fix dimension of dummy input

Cyrilvallez committed:
Add exception for prophetnet in _supports_num_logits_to_keep()

Cyrilvallez committed:
Update _supports_num_logits_to_keep() to use inspect.signature()

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -1446,6 +1446,13 @@ def _supports_default_dynamic_cache(self) -> bool:
         """
         return self._supports_cache_class and "jamba" not in self.__class__.__name__.lower()
 
+    def _supports_num_logits_to_keep(self) -> bool:
+        """
+        Return True if the current model supports the keyword argument `num_logits_to_keep` in forward()
+        to save memory. Checking it in this way allows to avoid using a new model attribute.
+        """
+        return "num_logits_to_keep" in set(inspect.signature(self.forward).parameters.keys())
```
Mmmmm I don't think we will add it to the signature, let's rather wait for this PR to be merged #31446 which will add kwargs

Cyrilvallez replied:
I see the related PR only adding kwargs to the forward of the attention blocks, not to the forward of the main ModelForCausalLM. Will it be added in the future or am I missing something?

ArthurZucker replied:
Ignore what I said, as this is high impact feature, we can add it to the signature

ArthurZucker reviewed (commented):
Thanks! Looks quite alright IMO!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31292). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -1202,11 +1208,16 @@ def forward(
         hidden_states = outputs[0]
-        logits = self.lm_head(hidden_states)
-        logits = logits.float()
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        if num_logits_to_keep is None:
+            logits = self.lm_head(hidden_states)
+        else:
+            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :])
```
Note on BC: despite [torch's type promotion](https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc), this change can be breaking on external scripts. For instance, if a user decides to add a FP16 tensor to `logits`, previously it would accumulate in FP32 and now it can accumulate in FP16.

Because of this, I'd suggest going through a deprecation cycle, e.g.
```py
        if labels is None:
            logger.warning_once(
                'Starting from v4.44, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)'
            )
        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
        if num_logits_to_keep is None:
            logits = self.lm_head(hidden_states).float()
        else:
            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
```

Then on v4.44 we would remove the `.float()`, giving the users time to adapt :)

gante reviewed (commented):
LGTM, thank you for further reducing the memory needs @Cyrilvallez 💛 

`num_logits_to_keep` is not the prettiest interface, but I can't think of a better one (as discussed in the PR that introduced it). 

I'm happy with the PR with the exception of BC handling

gante commented:
btw, a ratio of 3x lower peak memory consumption is 🔥 🔥 🔥

Cyrilvallez committed:
Add deprecation cycle + remove modification with pretraining_tp

Cyrilvallez committed:
Apply style

gante reviewed (approved):
LGTM 👍

Cyrilvallez committed:
Add most used models

Cyrilvallez committed:
Apply style

Cyrilvallez commented:
I just added the change to more models and rebased to avoid conflicts with new commits in main!
For Cohere-based models, I most notably computed a **memory gain ratio of 6.68** due to the very large 256k vocabulary size 🚀🔥

ArthurZucker commented on src/transformers/models/cohere/modeling_cohere.py:
```diff
@@ -1126,12 +1132,22 @@ def forward(
+            )
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        # TODO: remove those 2 float() operations in v4.44
+        if num_logits_to_keep is None:
+            logits = self.lm_head(hidden_states).float()
+        else:
+            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
```
we can default num_logits to keep to 0 to always slice (no codepaths)

ArthurZucker reviewed (commented):
Overall LGTM

ArthurZucker commented:
Make sure to rebase as the state of the main branch was changed quite a bit!

Cyrilvallez committed:
Make `num_logits_to_keep` an int in all cases to remove if-else clause

Cyrilvallez commented:
DO NOT MERGE YET
Everything else is good, but still need to sort out the logger.warning_once/compile issue

Cyrilvallez committed:
Add compile check for the warning

Cyrilvallez commented:
@ArthurZucker @gante everything is now ready.
From my tests, it seems like `compile` does not support any print-like functionality at the moment, either from `print`, `logger` or `warnings`.
I first wanted to add a `logger.warning_once_compile_safe` function which I thought would simplify things and come in handy in the future as well, but couldn't because it needs to import `torch` in the `logging` module which breaks things. 
So I just added a compile check everywhere.

Cyrilvallez committed:
Fix torch versions

Cyrilvallez committed:
style

ringohoffman commented:
@ArthurZucker is this planned for review this week? I’m pretty eager to consume this PR.

ArthurZucker commented:
Yes!

Oxi84 commented:
Looking forward to testing this out, gemma2 uses a lot of memory  otherwise and is a top model.

ArthurZucker commented on src/transformers/models/cohere/modeling_cohere.py:
```diff
@@ -1021,12 +1028,19 @@ def forward(
-        logits = self.lm_head(hidden_states)
+        if labels is None and not is_torchdynamo_compiling():
+            logger.warning_once(
+                "Starting from v4.44, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```
```suggestion
                "Starting from v4.45, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```

ArthurZucker commented on src/transformers/generation/utils.py:
Maybe let's comment that we need .float() for full precision soft

ArthurZucker reviewed (approved):
Well LGTM one thing missing is a test in the mixins !

ringohoffman commented:
Hey @Cyrilvallez, thanks for your work. Just checking in regarding this PR. Do you have a plan to finish it up some time soon?  I'm very excited for it to land!

Cyrilvallez commented:
Hi @ringohoffman, don't worry I am not forgetting about this 😉 I'm currently on vacation so I will try to wrap it up quickly end of August when I come back if I have time. Worst case scenario, it will be ready mid-September.

In the meantime, you can install transformers from my fork if you want to already benefit from it (pip install git+https://github.com/Cyrilvallez/transformers@logits-dtype). Or even better, you can clone my fork and rebase it on transformers/main to get all the new stuff + this PR.

Boubou78000 commented:
Does this PR actually fixes gemma2 or just Gemma?

Cyrilvallez commented:
Gemma2 was not released yet when I started this, but don't worry I will add it as well, it's on the roadmap 🤗

Cyrilvallez committed:
Add gemma2

Cyrilvallez committed:
Update warning version

Cyrilvallez committed:
Add comment about .float operations in generation utils

Cyrilvallez committed:
Add tests in GenerationTesterMixin and ModelTesterMixin

Cyrilvallez committed:
Fix batch size for assisted decoding in tests

Cyrilvallez committed:
fix small issues in test

Cyrilvallez committed:
refacor test

Cyrilvallez committed:
fix slicing removing dim issue

Cyrilvallez commented:
@ArthurZucker I added support for Gemma2 as well as tests, ready for last check 🤗
Red CIs are not related to the PR

ArthurZucker commented:
Can you rebase to fix the CI? 🤗

Cyrilvallez committed:
Add nemotron support (should fix check-copy issue in CIs)

Cyrilvallez committed:
Trigger new CIs

Cyrilvallez commented:
Fixed it!
Last red CI is the `libtorch_cuda.so: cannot open shared object file: No such file or directory` issue that was also present on recent PRs.

gante commented:
@Cyrilvallez the CI images built after [this PR](https://github.com/huggingface/transformers/pull/32918) gets merged should fix the CI issue we're seeing 🤗

Cyrilvallez committed:
Trigger new CIs

Cyrilvallez commented:
Indeed it worked, @ArthurZucker everything is now green!

ArthurZucker commented on src/transformers/models/cohere/modeling_cohere.py:
```diff
@@ -1071,12 +1078,19 @@ def forward(
-        logits = self.lm_head(hidden_states)
+        if labels is None and not is_torchdynamo_compiling():
+            logger.warning_once(
+                "Starting from v4.45, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```
```suggestion
                "Starting from v4.46, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```

ArthurZucker replied:
should be changed everywhere as next version is 4.45!

Cyrilvallez replied:
Done but the `examples_tensorflow` test started to fail again on the CI 🥲

ArthurZucker reviewed (approved):
LGTM!

Cyrilvallez committed:
Bump version

Cyrilvallez committed:
Bump version in TODO

Cyrilvallez committed:
Trigger CIs

gante commented:
@Cyrilvallez rebasing (again) will get rid of the CI error

(bear with us, we're chasing these CI blockers 🙏 )

Cyrilvallez committed:
remove blank space

Cyrilvallez committed:
Trigger CIs

Cyrilvallez commented:
No worries! All good on the CIs and ready to be merged 🤗
